File: package/main/src/Map/groupByToMap.ts
/**
 * Groups elements of an array into a Map based on a given iteratee function.
 * Unlike groupBy which returns Record<K, T[]>, this preserves insertion order
 * and allows any key type (including objects and symbols).
 * @param array Array to group
 * @param iteratee Function to determine the group key for each element
 * @returns Map with grouped elements keyed by the iteratee result
 * @example
 * groupByToMap([6.1, 4.2, 6.3], Math.floor); // Map { 6 => [6.1, 6.3], 4 => [4.2] }
 * groupByToMap(["one", "two", "three"], (str) => str.length); // Map { 3 => ['one', 'two'], 5 => ['three'] }
 */
export const groupByToMap = <T, K>(
  array: T[],
  iteratee: (value: T, index: number, array: T[]) => K,
): Map<K, T[]> => {
  const result = new Map<K, T[]>();
  const length = array.length;
  for (let index = 0; index < length; index++) {
    const value = array[index];
    const key = iteratee(value, index, array);
    const bucket = result.get(key);
    if (bucket === undefined) {
      result.set(key, [value]);
    } else {
      bucket.push(value);
    }
  }
  return result;
};

File: package/umt_python/src/map/group_by_to_map.py
from collections.abc import Callable, Hashable
from typing import TypeVar

T = TypeVar("T")
K = TypeVar("K", bound=Hashable)


def group_by_to_map(
    array: list[T],
    iteratee: Callable[[T, int, list[T]], K],
) -> dict[K, list[T]]:
    """
    Groups elements of a list into a dict based on a given iteratee function.

    Unlike group_by which returns a dict[str | int, list[T]], this preserves
    insertion order and allows any hashable key type. Python's dict preserves
    insertion order (3.7+), mirroring the TS Map semantics.

    Args:
        array: List to group.
        iteratee: Function to determine the group key for each element.

    Returns:
        Dict with grouped elements keyed by the iteratee result.

    Example:
        >>> group_by_to_map([6.1, 4.2, 6.3], lambda x, i, arr: int(x))
        {6: [6.1, 6.3], 4: [4.2]}
        >>> group_by_to_map(["one", "two", "three"], lambda x, i, arr: len(x))
        {3: ['one', 'two'], 5: ['three']}
    """
    result: dict[K, list[T]] = {}
    for index, value in enumerate(array):
        key = iteratee(value, index, array)
        bucket = result.get(key)
        if bucket is None:
            result[key] = [value]
        else:
            bucket.append(value)
    return result
